Add analytics event for extract column in chill mode

romeovs opened the pull request:
Closes https://github.com/metabase/metabase/issues/41773

### Description

Adds an analytics event that fires when the custom column is created from the table mode via Extract button in the column header

romeovs commented:
Superseded by https://github.com/metabase/metabase/pull/41767

ranquild reviewed (changes requested):
`structEvents` are deprecated google analytics events that should not be used. Please use snowplow events.

romeovs committed:
Add trackColumnExtractViaHeader helper

romeovs committed:
Track column extractions via header

romeovs committed:
Add Lib.extract wrapper

romeovs committed:
Fix analytics event name

romeovs committed:
Add tests for extract action analytics

romeovs commented on frontend/src/metabase/querying/analytics.ts:
```diff
@@ -0,0 +1,40 @@
+};
+
+function expressionsUsedBy(tag: string) {
+  switch (tag) {
```
This is a bit fragile: when we add more extractions, we need to add the expressions used here too.

But AFAIC there is no easy way to get the expressions introduced by an extraction from the extraction itself.

ranquild replied:
We could make `ColumnExtractionInfo.tag` to be a enum and not just a string. We could also define a mapping type used here in a way to make it fail if there is a new tag that is not supported here. Like:
```
Record<ColumnExtractionTag, string>
```

Another way could to be to make analytic events match the expression tag either by updating the tag or the analytics spec.

romeovs replied:
Ok I've done exactly that, please have another look.

ranquild reviewed (commented):
Inline comments.

romeovs committed:
Set up extractions metadata helper

romeovs committed:
Use functionsUseByExtraction in analytics

romeovs committed:
Add custom type for ExtractionTag

romeovs committed:
Use ExtractionTag in analytics

ranquild commented:
@romeovs same as https://github.com/metabase/metabase/pull/42095#issuecomment-2088397865

romeovs committed:
Add extractionExpression wrapper

bshepherdson committed:
[MBQL lib] Add `Lib.columnExtractDrillExtractions`

This gets the extractions from a column-extract drill, which is needed
in the UI.

romeovs committed:
Add extractionsForDrill helper

romeovs committed:
Fix type signature of extractionExpression

romeovs committed:
Use expressionParts for getting expression function

romeovs committed:
Walk expression to get functions used

romeovs commented:
@ranquild please have another look, Braden added a helper so we could actually get the extractions from the drill.

That is used to get the functions from the expression generated by the extraction correctly.

romeovs committed:
Simplify ColumnExtractionTag

romeovs committed:
Merge branch 'master' into 41773-add-analytics-event-to-chill-mode-extract-column

romeovs committed:
Fix expected expression

ranquild commented on frontend/src/metabase-lib/extractions.ts:
```diff
@@ -9,10 +36,39 @@ export function columnExtractions(
+    if (!arg || !(typeof arg === "object")) {
+      return;
+    }
+    if ("operator" in arg) {
```
`if (`typeof arg === "object") walk(...)` without the first if

ranquild replied:
https://github.com/metabase/metabase/blob/4b71e9aa645117a944af409882bda33f64225bf9/frontend/src/metabase-lib/filter.ts#L570

ranquild replied:
because `typeof null === "object"`

ranquild replied:
feel free to move this utility to a different place

romeovs committed:
Be more explicit with what the guard does
